### DISCUSSION TITLE
Why TOML for configs?

### DISCUSSION THREAD

**Discussion by glyh** [helix-editor/helix#2277] [General] [answered]
As title. I'd like to know your ideas on this. I would personally prefer a full blown Programming language.

**Comment by workingj** [comment]
@glyh
I think one of the main reasons the devs have in mind is simplicity and ease of use.

**Reply by glyh** [reply]
But then you can't do a lot of things... What if a plugin requires configurations that need more than data? Say users can provide custom hooks?

**Comment by the-mikedavis** [comment]
TOML is currently used _because_ it's not a general purpose programming language. Currently config is just config. We're starting to reach the ends of this being sufficient though. More complicated configs like custom statuslines or rebinding entire minor modes are difficult to tune with just TOML without imposing some kind of hacky domain specific language on top of it.

At some point I imagine there will be a switch from TOML to something more flexible. I would really prefer it not be a general purpose programming language though. Something powerful enough to achieve what we need is great but I would find it really distasteful if you could (for example) spawn an HTTP server using a config language.

**Reply by glyh** [reply]
Nobody would spawn HTTP server with a config language... But I think sometimes it have to be Turing Complete.

**Reply by the-mikedavis** [reply]
That's meant to be hyperbole. The language's power is not a trivial aspect: it should be capable of expressing everything you need configuration-wise but it must not powerful enough to compete with the plugin system. For example, [dhall](https://github.com/dhall-lang/dhall-lang) is appropriately powerful. Python is not.

**Reply by kirawi** [reply]
To add, the current TOML configuration was (is?) intended to be a temporary solution.

**Reply by David-Else** [reply]
I really hope Helix does not slowly turn into Neovim. We already have Neovim. Having config being in TOML is a massive win for simplicity and usability, it is kind of ridiculous to have to use a programming language to configure an editor.

VS Code uses JSON for config and I have never heard anyone say they find it limiting, all the options people need are in the plugins themselves.

**Reply by glyh** [reply]
> I really hope Helix does not slowly turn into Neovim. We already have Neovim. Having config being in TOML is a massive win for simplicity and usability, it is kind of ridiculous to have to use a programming language to configure an editor.
> 
> VS Code uses JSON for config and I have never heard anyone say they find it limiting, all the options people need are in the plugins themselves.

So VSCode misses some functionality that could be possible in neovim. For example, in nvim-cmp, we have callbacks for Tab completion:
```lua
      ["<Tab>"] = cmp.mapping(function(fallback)
        local luasnip = require("luasnip")
        if cmp.visible() then
          cmp.select_next_item()
        elseif luasnip.expand_or_jumpable() then
          luasnip.expand_or_jump()
        elseif has_words_before() then
          cmp.complete()
        else
          fallback()
        end
      end, { "i", "s" }),
```
You can use which ever snippet plugin you like, but if you use something like JSON, such stuff must be hard coded into the plugin, and you have to expect that plugin maintainers care about your use case.

This might be a win for simplicity, but definitely not for usability.

**Comment by archseer** [comment, marked as answer]
It's a stopgap until we get something like https://github.com/helix-editor/helix/issues/122 implemented.

**Comment by theowenyoung** [comment]
VSCode also supports JSON config only, but it still has plenty of extensions and a good user experience.
